theories/newth/MapAux.ec

```diff
@@ -0,0 +1,146 @@
+(*---------------------- Auxiliary Lemmas on Maps ----------------------*)
+
+prover [""].
+
+require import AllCore SmtMap FSet StdOrder.
+import IntOrder.
+
+lemma get_none (m : ('a, 'b) fmap, x : 'a) :
+  x \notin m => m.[x] = None.
+proof. by rewrite domE. qed.
+
+lemma get_some (m : ('a, 'b) fmap, x : 'a) :
+  x \in m => m.[x] = Some (oget m.[x]).
+proof. move=> /domE; by case m.[x]. qed.
+
+lemma set_same (m : ('a, 'b) fmap, x : 'a) :
+  x \in m => m.[x <- oget m.[x]] = m.
+proof.
+move=> x_in_m.
+apply fmap_eqP => y.
+case (y = x) => [->> | ne_y_x].
+by rewrite get_set_sameE get_some.
+by rewrite get_setE ne_y_x.
+qed.
+
+lemma set_eq (m : ('a, 'b) fmap, x : 'a, y : 'b) :
+  m.[x] = Some y => m.[x <- y] = m.
+proof.
+move=> m_get_x_eq_y.
+have x_in_m : x \in m by rewrite domE m_get_x_eq_y.
+have -> : y = oget m.[x] by rewrite m_get_x_eq_y oget_some.
+by rewrite set_same.
+qed.
+
+lemma frng_set (m : ('a, 'b) fmap, x : 'a, y : 'b) :
+  frng m.[x <- y] = frng (rem m x) `|` fset1 y.
+proof.
+apply fsetP => z; rewrite in_fsetU in_fset1 2!mem_frng 2!rngE /=.
+split => [[x'] | [[x'] | ->]].
+case (x' = x) => [-> | ne_x'_x].
+by rewrite get_set_sameE /= => ->.
+rewrite get_setE ne_x'_x /= => get_x'_some_z.
+left; exists x'; by rewrite remE ne_x'_x.
+rewrite remE.
+case (x' = x) => // ne_x'_x get_x'_some_z.
+exists x'; by rewrite get_setE ne_x'_x.
+exists x; by rewrite get_set_sameE.
+qed.
+
+lemma eq_except_ne_in (x y : 'a, m1 m2 : ('a, 'b) fmap) :
+  eq_except (pred1 x) m1 m2 => y <> x =>
+  y \in m1 => y \in m2.
+proof.
+move=> /eq_exceptP @/pred1 eq_exc ne_y_x.
+by rewrite 2!domE eq_exc.
+qed.
+
+lemma eq_except_setr_as_l (m1 m2 : ('a, 'b) fmap) x:
+  x \in m1 => eq_except (pred1 x) m1 m2 =>
+  m1 = m2.[x <- oget m1.[x]].
+proof.
```

Comment:
Why `pred1` explicitly?

Reply:
This one *doesn't* hold for an arbitrary predicate. I think I stole this from the old NewFMap.